### PR TITLE
TLSv1.3: Verify that key change messages end on a record boundary

### PR DESCRIPTION
**Pull request by mattcaswell**
##### Checklist

- [x] tests are added or updated

##### Description of change

TLSv1.3 has 3 handshake messages (ServerHello, Finished and KeyUpdate) that signal a change in the keys being used. Because keys are changing it is important (and a requirement of the spec) that the end of those messages must coincide with the end of the record. Otherwise we could process data believing it have been decrypted with the new keys, but actually it was decrypted prior to the key change using the old keys. This also adds tests for the 3 key change scenarios.

In passing this fixes an SSL_has_pending() bug which did not recognise decrypted handshake data that has not yet been read as "pending".

The first commit is also for 1.1.0.

### CODE REVIEW

File: ssl/record/rec_layer_s3.c [hunk 1]
```diff
@@ -77,11 +77,28 @@ void RECORD_LAYER_release(RECORD_LAYER *rl)
...
 
+/* Checks if we have decrypted unread record data pending */
+int RECORD_LAYER_processed_read_pending(const RECORD_LAYER *rl)
+{
+    size_t num_recs, curr_rec;
+    const SSL3_RECORD *rr;
+
+    rr = rl->rrec;
+    num_recs = RECORD_LAYER_get_numrpipes(rl);
+    for (curr_rec = 0;
```
Comment (richsalz) [inline ssl/record/rec_layer_s3.c]: You could move the assignments up to the declaration.  Then change the for loop into a while.

Comment (mattcaswell) [inline ssl/record/rec_layer_s3.c]: Ok. Done.

**Comment by mattcaswell** [comment]
Updated with requested changes.

**Comment by mattcaswell** [comment]
Added a new commit to fix the travis failure (the number of tests to skip was wrong if TLSv1.3 was disabled). @richsalz please reconfirm.

**Comment by richsalz** [comment]
reconfirm

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**